ext/drreg/drreg.c

```diff
@@ -1599,19 +2411,66 @@ drreg_restore_app_aflags(void *drcontext, instrlist_t *ilist, instr_t *where)
  */
 
 static bool
-is_our_spill_or_restore(void *drcontext, instr_t *instr, bool *spill OUT,
-                        reg_id_t *reg_spilled OUT, uint *slot_out OUT, uint *offs_out OUT)
+is_our_spill_or_restore(void *drcontext, instr_t *instr, instr_t *next_instr,
+                        bool *spill OUT, reg_id_t *reg_spilled OUT, uint *slot_out OUT,
+                        uint *offs_out OUT, bool *is_indirectly_spilled OUT)
 {
     bool tls;
     uint slot, offs;
     reg_id_t reg;
-    if (!instr_is_reg_spill_or_restore(drcontext, instr, &tls, spill, &reg, &offs))
+    bool is_spilled;  /* Flag denoting spill or restore*/
+    bool is_indirect; /* Flag denoting direct or indirect access */
+
+    is_indirect = false;
+    if (is_indirectly_spilled != NULL)
+        *is_indirectly_spilled = is_indirect;
+
+    if (!instr_is_reg_spill_or_restore(drcontext, instr, &tls, &is_spilled, &reg, &offs))
         return false;
-    /* Is this from our raw TLS? */
+    /* Is this from our direct raw TLS for gpr registers ? */
     if (tls && offs >= tls_slot_offs &&
         offs < (tls_slot_offs + ops.num_spill_slots * sizeof(reg_t))) {
         slot = (offs - tls_slot_offs) / sizeof(reg_t);
-    } else {
+    }
+#ifdef X86
+    else if (tls && offs == tls_simd_offs &&
+             !(is_spilled) /* Cant be a spill bc loading block */) {
+        /* In order to detect indirect spills, the loading of the pointer
+         * to the indrect block must be done exactly prior. We assume that
+         * nobody else can interfere with our indirect load sequence for
+         * simd registers.
+         */
+        ASSERT(next_instr != NULL, "next_instr cannot be NULL");
+        /* FIXME i#3844: Might need to change this assert when
+         * supporting other register spillage.
+         */
+        ASSERT(instr_get_opcode(next_instr) == OP_movdqa ||
+                   instr_get_opcode(next_instr) == OP_vmovdqa,
+               "next instruction needs to be a mov");
+        is_indirect = true;
+        opnd_t dst = instr_get_dst(next_instr, 0);
+        opnd_t src = instr_get_src(next_instr, 0);
+
+        if (opnd_is_reg(dst) && reg_is_vector_simd(opnd_get_reg(dst)) &&
+            opnd_is_base_disp(src)) {
+            reg = opnd_get_reg(dst);
+            is_spilled = false;
+            int disp = opnd_get_disp(src);
+            slot = disp / REG_SIMD_SIZE;
+        } else if (opnd_is_reg(src) && reg_is_vector_simd(opnd_get_reg(src)) &&
+                   opnd_is_base_disp(dst)) {
+            reg = opnd_get_reg(src);
+            is_spilled = true;
+            int disp = opnd_get_disp(dst);
+            /* Displacement spans over SIMD sizes. Perform division to get slot */
```

Comment:
I would also re-phrase, isn't the slot of size simd size? The displacement is just a memory offset. Maybe s/Displacement spans/Slots span/?

Reply:
Fly-by comment on style: please follow the style guide which requests punctuation and complete sentences for multi-sentence comments.

Reply:
Done.